Added schema and tests for LogStreams

mcalster opened the pull request:
<!--- 

**IMPORTANT:** Please submit pull requests to [alexkappa/terraform-provider-auth0](https://github.com/alexkappa/terraform-provider-auth0). This helps maintainers organize work more efficiently.

See what makes a good Pull Request at : https://github.com/alexkappa/terraform-provider-auth0/blob/master/.github/CONTRIBUTING.md#pull-requests 

--->
### Proposed Changes
				resource "auth0_log_stream" "my_log_stream" {
					name = "Acceptance-Test-LogStream-http"
					type = "http"
					sink { 
					  http_endpoint = "https://example.com/webhook/logs"
					  http_content_type = "application/json"
					  http_content_format = "JSONLINES"
					  http_authorization = "AKIAXXXXXXXXXXXXXXXX"
					}
				}
* Add new LogStream Resource
* Update auth0 client version.

Fixes #264 
Blocked by https://github.com/go-auth0/auth0/pull/144

#### Acceptance Test Output

```
$ make testacc TESTS=TestAccLogStream*
==> Checking that code complies with gofmt requirements...
?       github.com/alexkappa/terraform-provider-auth0   [no test files]
=== RUN   TestAccLogStreamHttp
--- PASS: TestAccLogStreamHttp (2.60s)
=== RUN   TestAccLogStreamEventBridge
--- PASS: TestAccLogStreamEventBridge (3.43s)
=== RUN   TestAccLogStreamEventGrid
    resource_auth0_log_stream_test.go:100: this test requires an active subscription
--- SKIP: TestAccLogStreamEventGrid (0.00s)
=== RUN   TestAccLogStreamDatadog
--- PASS: TestAccLogStreamDatadog (1.50s)
=== RUN   TestAccLogStreamSplunk
--- PASS: TestAccLogStreamSplunk (1.80s)
PASS
coverage: 10.2% of statements
ok      github.com/alexkappa/terraform-provider-auth0/auth0     7.991s  coverage: 10.2% of statements
?       github.com/alexkappa/terraform-provider-auth0/auth0/internal/debug      [no test files]
testing: warning: no tests to run
PASS
coverage: 0.0% of statements
ok      github.com/alexkappa/terraform-provider-auth0/auth0/internal/random     0.222s  coverage: 0.0% of statements [no tests to run]
testing: warning: no tests to run
PASS
coverage: 0.0% of statements
ok      github.com/alexkappa/terraform-provider-auth0/auth0/internal/validation 0.108s  coverage: 0.0% of statements [no tests to run]
?       github.com/alexkappa/terraform-provider-auth0/version   [no test files]
```

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

mcalster commented:
@alexkappa I think this is good to go

kpurdon commented:
Just came here looking for this feature. Things look stalled from a review standpoint. Any update @alexkappa, thanks for your work!

abulford commented on auth0/resource_auth0_log_stream.go:
```diff
@@ -0,0 +1,330 @@
+	o := &management.EventBridgeSink{
+		AWSAccountID:          String(d, "aws_account_id"),
+		AWSRegion:             String(d, "aws_region"),
+		AWSPartnerEventSource: String(d, "aws_partner-event_source"),
```
Is the hyphen intentional?  On line 217 you've used `aws_partner_event_source` but here you've got `aws_partner-event_source`.

mcalster replied:
Good catch, I've fixed it.

abulford replied:
👍

abulford commented:
@mcalster thanks for this, I'm also interested in getting this functionality in the provider.

I've tried this locally and it seems to work well for me!  I'm using the Amazon EventBridge type, so can't comment on the other types.  The build I used was slightly different from this branch because I merged your [OAuth2 PR](https://github.com/alexkappa/terraform-provider-auth0/pull/272) into this branch locally, and updated your [auth0 PR](https://github.com/go-auth0/auth0/pull/144) with `v4.7.0` locally, because I also need the OAuth2 connection functionality.  Those merges went fine with no conflicts.

I've tried creating, updating (by changing the AWS region, which actually results in a re-create), deleting and creating again and all looks good.

The only strange thing I noticed is that when I modified the region it told me that the `http_custom_headers` and `splunk_secure` properties would be deleted, but I never had those set:
```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # auth0_log_stream.log_stream must be replaced
-/+ resource "auth0_log_stream" "log_stream" {
      ~ aws_partner_event_source = (sensitive value)
      ~ id                       = "<REDACTED>" -> (known after apply)
        name                     = "Auth0Events"
      ~ status                   = "active" -> (known after apply)
        type                     = "eventbridge"

      ~ sink {
            aws_account_id      = (sensitive value)
          ~ aws_region          = (sensitive value)
          - http_custom_headers = [] -> null
          - splunk_secure       = false -> null
        }
    }

Plan: 1 to add, 0 to change, 1 to destroy.
```

It didn't cause any problems but maybe means something isn't quite set up correctly?

Also, I need the `aws_partner_event_source` property as an output so I can use it to set up the event bus in AWS.  Locally I've added this as a top-level computed property which returns the sink's `AWSPartnerEventSource` property if the sink is of type `*management.EventBridgeSink`, or `nil` otherwise.  This seems to work for me though I've not tried it with other types.

I'm not sure what best to do with this - I could submit it as a PR to your PR, or wait until this gets merged then create a new PR in this repo.  Or if you're happy to add it into this PR yourself that's great too, I can send you a diff or something if you like.

abulford commented on auth0/resource_auth0_log_stream.go:
```diff
@@ -0,0 +1,330 @@
+						"aws_region": {
+							Type:      schema.TypeString,
+							Optional:  true,
+							Sensitive: true,
```
Does this need to be marked as sensitive?  Same with `aws_account_id`.

mcalster replied:
Region shouldn't be sensitive. AWS Account Id and Azure Subscription Id are generally considered secret.

abulford replied:
👍

mcalster commented:
@abulford Thanks for looking into it! I've only tested it with Azure Event grid.  You are more than welcome to PR my PR.  I will look at it and merge it ASAP. (propbably tonight or over the weekend).
It will make it much faster for me :-)

abulford commented:
@mcalster no problem, thanks for creating it!  OK, I'll try to get that done today.

If you're using Azure, would you like me to add in the `azure_partner_topic` property as well?  Would that be considered sensitive?

mcalster commented:
@abulford let me know if you are close with a PR. Otherwise I'll probably look at it tonight

abulford commented:
@mcalster sorry, I didn't get to this last night but submitted the PR now.

There were lots of different ways I could have done this and I'm not sure what I went with was necessarily best, so feel free to merge and change as you see fit, or if you let me know what you think and I can make modifications within the PR.

Thanks!

mcalster commented:
@abulford  thx for the PR, it was much easier for me to get going. I copied some of your changes instead of merging the PR.

The azure_partner_topic and aws_partner_event_source were already mapped in the specific flatten methods, so I didn't move those. They were missing in the shcema though, so I took used those parts. Could I ask you to take another round of review?

mcalster commented:
> The only strange thing I noticed is that when I modified the region it told me that the `http_custom_headers` and `splunk_secure` properties would be deleted, but I never had those set:

I can't see how this can happen, can you reproduce with the latest version of this PR?

abulford commented:
@mcalster thanks for incorporating some of my changes.

Sorry, I should have explained the rationale for my change better.  I did initially just expose the existing properties by including them in the schema, but found that I then had to access the property in my Terraform code like this:
```
auth0_log_stream.log_stream.sink[0].aws_partner_event_source
```
I wasn't particularly keen on the need to access the first element of the list of sinks, and generally it didn't feel consistent with how I'm used to accessing properties in other resources.  Given that this resource represents a single log stream in Auth0, and there's no option to provide multiple sinks within a single log stream resource, it felt appropriate to expose these properties as top level elements which will be set depending on the type specified.

So I instead declared the properties at the top level of the resource, which meant I couldn't take advantage of the work already done in the flatten and expand functions, but means I can access the properties directly in my Terraform code, like this:
```
auth0_log_stream.log_stream.aws_partner_event_source
```
This is how I'd instinctively expect to be able to access this property, though maybe others would expect something different.

Thinking about it, might it make sense to do this with all the sink properties...?  Currently the Terraform resource directly reflects the structure of the data used by the Auth0 management API, but does it need to?  It might make the resource easier to use if all properties were at the top level, with properties conflicting with each other so only valid combinations can be set.

I don't really mind which way this is implemented and am happy to update my Terraform code to match, but wanted to explain my rationale in case you hadn't realised how it would mean the property needs to be accessed.  What do you think?

abulford commented:
> > The only strange thing I noticed is that when I modified the region it told me that the `http_custom_headers` and `splunk_secure` properties would be deleted, but I never had those set:
> 
> I can't see how this can happen, can you reproduce with the latest version of this PR?

I'm afraid the issue still exists, and actually causes more of a problem than I'd realised.  Previously I tested updates by modifying the AWS region, which actually resulted in a re-creation of the resource.  This time I modified the name which allowed the provider to attempt an update rather than a re-creation, which fails.

So, the plan shows:
```
Terraform will perform the following actions:

  # auth0_log_stream.log_stream will be updated in-place
  ~ resource "auth0_log_stream" "log_stream" {
        id     = "<REDACTED>"
      ~ name   = "Auth0Events" -> "Auth0Events1"
        status = "active"
        type   = "eventbridge"

        sink {
            aws_account_id           = (sensitive value)
            aws_partner_event_source = "<REDACTED>"
            aws_region               = "eu-west-1"
            http_custom_headers      = []
            splunk_secure            = false
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```
When attempting to apply the update, it fails:
```
Error: 400 Bad Request: Payload validation error: 'Data does not match any schemas from 'oneOf'' on property sink.

  on auth0_log_stream.tf line 1, in resource "auth0_log_stream" "log_stream":
   1: resource "auth0_log_stream" "log_stream" {
```
Which I expect is down to the unexpected properties being present.

abulford commented on auth0/resource_auth0_log_stream.go:
```diff
@@ -0,0 +1,341 @@
+							Optional: true,
+							ForceNew: true,
+						},
+						"aws_partner_event_source": {
```
I'm not sure how much of a security risk exposure of this value would be,  but I'd err on the side of this being marked as sensitive, and I would guess the same would apply to the Azure equivalent.

abulford commented on go.mod:
```diff
@@ -8,5 +8,7 @@ require (
+	gopkg.in/auth0.v4 v4.7.0
 )
+
+replace gopkg.in/auth0.v4 => /Users/kagli/code/auth0
```
Not sure if you intended to keep this in and remove once the PR has been approved?  Thought I should flag it in case not!

mcalster commented:
> I'm afraid the issue still exists, and actually causes more of a problem than I'd realised.

I've digged a bit deeper, and it could be caused by a bug in the PR I made to the auth0 sdk. @alexkappa made a refactoring to fix it when it was merged into version 5.2.2. Basically the default initialized values wasn't correctly managed. 

The fact that LogStream support is added in v5 of the sdk, maybe we should make this PR work with v5 going forward?

mcalster commented:
> I don't really mind which way this is implemented and am happy to update my Terraform code to match,

Thx for clarifying. I didn't catch the correct intention in your changes, but I totally agree. It isn't a good experience to get the values via an index. It annoys me every time I have to do it, so let's agree on not introducing it here as well 😋So I guess the cleanest way it to move all the sink parameters up to the resource level? I think having some of the parameters in a sink config and other sink attributes at the resource level seems messy?

kpurdon commented:
Sorry to jump in here again ...

> The fact that LogStream support is added in v5 of the sdk, maybe we should make this PR work with v5 going forward?

I too have a PR that needs v5 so I opened PR https://github.com/alexkappa/terraform-provider-auth0/pull/291 to upgrade this provider from v4 -> v5 of go-auth0.

abulford commented:
>The fact that LogStream support is added in v5 of the sdk, maybe we should make this PR work with v5 going forward?

@mcalster I hadn't realised it had been merged in, v5 sounds like a plan.  Good to see you've got a PR already, @kpurdon!

>So I guess the cleanest way it to move all the sink parameters up to the resource level? I think having some of the parameters in a sink config and other sink attributes at the resource level seems messy?

I think having them all at the top level probably makes most sense, yeah.

I can have a go at implementing this tomorrow evening (GMT), though if you get there first I'm happy to give it a test.  Ideally this would all be on top of the v5 upgrade, so I'm not sure if it's better to wait for that to be merged, or to rebase the branch off of #291?

alexkappa commented:
Hi everyone, thank you for your efforts into getting log streams into the provider. And my apologies for the delay in answering or reviewing your work.

I will give this my full attention and help merge this as soon as we can.

alexkappa commented on auth0/resource_auth0_log_stream.go:
```diff
@@ -0,0 +1,341 @@
+		case management.LogStreamSinkSplunk:
+			c.Sink = expandLogStreamSplunkSink(d)
+		default:
+			log.Printf("[WARN]: Raise an issue with the auth0 provider in order to support it:")
+			log.Printf("[WARN]: 	https://github.com/alexkappa/terraform-provider-auth0/issues/new")
```
```suggestion
			log.Printf("[WARN]: Unsupported log stream sink %s", s)
			log.Printf("[WARN]: Raise an issue with the auth0 provider in order to support it:")
			log.Printf("[WARN]: 	https://github.com/alexkappa/terraform-provider-auth0/issues/new")
```

alexkappa commented on auth0/resource_auth0_log_stream_test.go:
```diff
@@ -0,0 +1,192 @@
+		Steps: []resource.TestStep{
+			{
+				Config: `
+				resource "auth0_log_stream" "my_log_stream" {
+					name = "Acceptance-Test-LogStream-aws"
+					type = "eventbridge"
+					sink {
+						aws_account_id = "999999999999"
+						aws_region = "us-west-2"
+					}
+				}
```
We would preferably follow the convention here and define them as variables. Should also add randomization also so we can be able to run tests in parallel.

alexkappa commented on auth0/resource_auth0_log_stream_test.go:
```diff
@@ -0,0 +1,192 @@
+}
+
+//This test fails it subscription key is not valid, or Eventgrid Resource Provider is not registered in the subscription
+/*func TestAccLogStreamEventGrid(t *testing.T) {
```
It's best to skip this test instead of commenting it out.

alexkappa reviewed (approved):
@mcalster the PR looks in really good shape, following the conventions found in other places in the provider. I don't see any issues other than a few small changes.

I will head over to #291 to merge it in, and come back here and help finish up.

mcalster commented:
@abulford I've made the changes we discussed :-)  could you review again and verify that it is what we agreed? Please let me know if I need to do something in order to move to the v5 of the sdk.

alexkappa commented:
@mcalster I've merged the v5 PR so we should be able to proceed here.

I don't think I agree with @abulford on flattening sink values to the top level. While I do understand the clunkiness of having to access outputs using an index, it's the way we do things in other resources and I would prefer to remain aligned with that. Take the connection resource for example which this PR is largely based on. 

Unfortunately terraform doesn't offer an alternative to nesting other than a `TypeList` or `TypeSet` data structures.

abulford commented:
Ack, sorry to have thrown a spanner in the works 😞 .  For what it's worth, I prefer your latest version, @mcalster, as it's more consistent with other providers I'm used to and provides (IMO) a more natural way to use the resource, but appreciate @alexkappa's point that it makes it less consistent with the other resources already in the provider.

I'm afraid I've been completely waylaid with other work so I'm sorry for not implementing the flattening myself and for not testing this yet.  I don't think I'll manage to test it this evening but should be able to do so tomorrow morning.

mcalster commented:
> @mcalster I've merged the v5 PR so we should be able to proceed here.
> 
> I don't think I agree with @abulford on flattening sink values to the top level. While I do understand the clunkiness of having to access outputs using an index, it's the way we do things in other resources and I would prefer to remain aligned with that. Take the connection resource for example which this PR is largely based on.
> 
> Unfortunately terraform doesn't offer an alternative to nesting other than a `TypeList` or `TypeSet` data structures.

I've merged the changes from master and updated the PR to use v5. Reverted back to using Sink as a construct in HCL.

abulford commented:
@mcalster I've merged master into the feature branch and tested in its current (flattened) state, I've then reverted the flattening to bring it back to the nested structure and tested that.  Both versions generally work though there are still some niggles - unfortunately the v5 library didn't resolve the issue I found before.

So with the nested version of the provider I can initially create a log stream with no problems:
```
  # auth0_log_stream.log_stream will be created
  + resource "auth0_log_stream" "log_stream" {
      + id     = (known after apply)
      + name   = "Auth0Events"
      + status = (known after apply)
      + type   = "eventbridge"

      + sink {
          + aws_account_id           = (sensitive value)
          + aws_partner_event_source = (known after apply)
          + aws_region               = (sensitive value)
          + azure_partner_topic      = (known after apply)
        }
    }
```
If I then tweak the name to cause an update, the `http_custom_headers` and `splunk_secure` properties now appear:
```
  # auth0_log_stream.log_stream will be updated in-place
  ~ resource "auth0_log_stream" "log_stream" {
        id     = "<REDACTED>"
      ~ name   = "Auth0Events" -> "Auth0Events1"
        status = "active"
        type   = "eventbridge"

        sink {
            aws_account_id           = (sensitive value)
            aws_partner_event_source = "<REDACTED>"
            aws_region               = (sensitive value)
            http_custom_headers      = []
            splunk_secure            = false
        }
    }
```
When attempting to apply I get the following error:
```
Error: 400 Bad Request: Payload validation error: 'Data does not match any schemas from 'oneOf'' on property sink.

  on auth0_log_stream.tf line 1, in resource "auth0_log_stream" "log_stream":
   1: resource "auth0_log_stream" "log_stream" {
```
I instead change the region, to force a recreate, which I believe worked before:
```
  # auth0_log_stream.log_stream must be replaced
-/+ resource "auth0_log_stream" "log_stream" {
      ~ id     = "<REDACTED>" -> (known after apply)
        name   = "Auth0Events"
      ~ status = "active" -> (known after apply)
        type   = "eventbridge"

      ~ sink {
            aws_account_id           = (sensitive value)
          ~ aws_partner_event_source = "<REDACTED>" -> (known after apply)
          ~ aws_region               = (sensitive value)
          + azure_partner_topic      = (known after apply)
          - http_custom_headers      = [] -> null
          - splunk_secure            = false -> null
        }
    }
```
This succeeds in deleting the log stream, but creating the new one fails, with the same error:
```
Error: 400 Bad Request: Payload validation error: 'Data does not match any schemas from 'oneOf''.

  on auth0_log_stream.tf line 1, in resource "auth0_log_stream" "log_stream":
   1: resource "auth0_log_stream" "log_stream" {
```

A subsequent apply works, seemingly because by this point Terraform knows that the resource has been deleted, and starts from scratch in creating it again.

Removing the log stream in my Terraform code results in a successful deletion, though you can again see that those properties are unexpectedly set:
```
  # auth0_log_stream.log_stream will be destroyed
  - resource "auth0_log_stream" "log_stream" {
      - id     = "lst_0000000000001089" -> null
      - name   = "Auth0Events" -> null
      - status = "active" -> null
      - type   = "eventbridge" -> null

      - sink {
          - aws_account_id           = (sensitive value)
          - aws_partner_event_source = "aws.partner/auth0.com/redmatter-dev01-4da38e92-1c08-4277-9253-c92ebe093c6c/auth0.logs" -> null
          - aws_region               = (sensitive value)
          - http_custom_headers      = [] -> null
          - splunk_secure            = false -> null
        }
    }
```

Switching to the flattened version of the provider, creation works:
```
  # auth0_log_stream.log_stream will be created
  + resource "auth0_log_stream" "log_stream" {
      + aws_account_id           = (sensitive value)
      + aws_partner_event_source = (known after apply)
      + aws_region               = (sensitive value)
      + azure_partner_topic      = (known after apply)
      + id                       = (known after apply)
      + name                     = "Auth0Events"
      + status                   = (known after apply)
      + type                     = "eventbridge"
    }
```
And interestingly, the update doesn't show those extra properties:
```
  # auth0_log_stream.log_stream will be updated in-place
  ~ resource "auth0_log_stream" "log_stream" {
        aws_account_id           = (sensitive value)
        aws_partner_event_source = "<REDACTED>"
        aws_region               = (sensitive value)
        id                       = "<REDACTED>"
      ~ name                     = "Auth0Events" -> "Auth0Events1"
        status                   = "active"
        type                     = "eventbridge"
    }
```
But on applying, it does still fail in the same way:
```
Error: 400 Bad Request: Payload validation error: 'Data does not match any schemas from 'oneOf'' on property sink.

  on auth0_log_stream.tf line 11, in resource "auth0_log_stream" "log_stream":
  11: resource "auth0_log_stream" "log_stream" {
```

Changing the region to force a recreate also still fails, again without showing those extra properties:
```
  # auth0_log_stream.log_stream must be replaced
-/+ resource "auth0_log_stream" "log_stream" {
        aws_account_id           = (sensitive value)
      ~ aws_partner_event_source = "aws.partner/auth0.com/redmatter-dev01-6c98f657-24ad-4d72-86cf-d651537c1650/auth0.logs" -> (known after apply)
      ~ aws_region               = (sensitive value)
      + azure_partner_topic      = (known after apply)
      ~ id                       = "lst_0000000000001090" -> (known after apply)
        name                     = "Auth0Events"
      ~ status                   = "active" -> (known after apply)
        type                     = "eventbridge"
    }

Plan: 1 to add, 0 to change, 1 to destroy.
...
Error: 400 Bad Request: Payload validation error: 'Data does not match any schemas from 'oneOf''.

  on auth0_log_stream.tf line 11, in resource "auth0_log_stream" "log_stream":
  11: resource "auth0_log_stream" "log_stream" {
```
Though strangely in this case, when attempting to apply again, even though I would have thought Terraform would not think the resource has gone and start again, it fails again on creation:
```
  # auth0_log_stream.log_stream will be created
  + resource "auth0_log_stream" "log_stream" {
      + aws_account_id           = (sensitive value)
      + aws_partner_event_source = (known after apply)
      + aws_region               = (sensitive value)
      + azure_partner_topic      = (known after apply)
      + id                       = (known after apply)
      + name                     = "Auth0Events"
      + status                   = (known after apply)
      + type                     = "eventbridge"
    }
...
Error: 400 Bad Request: Payload validation error: 'Data does not match any schemas from 'oneOf''.

  on auth0_log_stream.tf line 11, in resource "auth0_log_stream" "log_stream":
  11: resource "auth0_log_stream" "log_stream" {
```
The only way to get it to all to work is to delete the resource in the Terraform code, apply, then reinstate the code and apply again.

I had a quick look and also couldn't see why this should be happening, though I expect it's something to do with the fact that these two properties are not strings like all the others.  Are they somewhere being set with `[]` and `false` instead of `null`?  I can have a closer look but wanted to get this done first.


One other small thing is that adding the computed values has had a slightly unexpected side-effect:
```
Terraform will perform the following actions:

  # auth0_log_stream.log_stream will be created
  + resource "auth0_log_stream" "log_stream" {
      + id     = (known after apply)
      + name   = "Auth0Events"
      + status = (known after apply)
      + type   = "eventbridge"

      + sink {
          + aws_account_id           = (sensitive value)
          + aws_partner_event_source = (known after apply)
          + aws_region               = (sensitive value)
          + azure_partner_topic      = (known after apply)
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```
It's a little bit confusing that both `aws_partner_event_source` and `azure_partner_topic` show up even though only one of them should be set in the end, but I suppose that's a consequence of making these computed values available.

Sorry for the long comment and for seemingly being a bit too late with the PRs - I've just noticed it looks like you've also done the v5 merge.  I can try testing that as well if that's of any use.

abulford commented:
I've built and tested your latest version and it looks good, though still with the update issue described above.

Re-creation is fine, though.  I made a mistake when testing - instead of changing to a different region (e.g. from `eu-west-1` to `us-east-1`) I just added a `1` to the end of the region name to force a change, as I had done with the name.  Obviously Auth0 didn't accept that, which caused the failure.  Using a valid region name allows the re-creation to work properly.

This reminds me of the sensitive properties, though - it does make it harder to see what's going on when properties are redacted like that.  I'm not sure if the region needs to be sensitive, though the `aws_partner_event_source` at least looks like something that perhaps should be.

mcalster commented:
> I've built and tested your latest version and it looks good, though still with the update issue described above.
> 
> Re-creation is fine, though. I made a mistake when testing - instead of changing to a different region (e.g. from `eu-west-1` to `us-east-1`) I just added a `1` to the end of the region name to force a change, as I had done with the name. Obviously Auth0 didn't accept that, which caused the failure. Using a valid region name allows the re-creation to work properly.
> 
> This reminds me of the sensitive properties, though - it does make it harder to see what's going on when properties are redacted like that. I'm not sure if the region needs to be sensitive, though the `aws_partner_event_source` at least looks like something that perhaps should be.

I've changed the sensitivity as requested :-)

mcalster commented:
@abulford  I've updated the test cases and it seems like I can reproduce your issue when changing the name of the log stream.
Running with debug logging I get the following warning:
2020/11/13 20:49:46 [WARN] Provider "auth0" produced an unexpected new value for auth0_log_stream.my_log_stream, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .sink[0].splunk_secure: was null, but now cty.False
      - .sink[0].splunk_token: inconsistent values for sensitive attribute
      - .sink[0].splunk_domain: was null, but now cty.StringVal("")
      - .sink[0].azure_resource_group: was null, but now cty.StringVal("")
      - .sink[0].azure_subscription_id: inconsistent values for sensitive attribute
      - .sink[0].datadog_region: inconsistent values for sensitive attribute
      - .sink[0].datadog_api_key: inconsistent values for sensitive attribute
      - .sink[0].splunk_port: was null, but now cty.StringVal("")
      - .sink[0].aws_account_id: inconsistent values for sensitive attribute
      - .sink[0].aws_region: was null, but now cty.StringVal("")
      - .sink[0].azure_region: was null, but now cty.StringVal("")

I think the culprit is splunk_secure, but I'm unsure how to fix it :-/

alexkappa commented:
> I think the culprit is splunk_secure, but I'm unsure how to fix it :-/

@mcalster I'm not sure if this is the same issue as the one @abulford has reported, but I've found that the `splunk_secure` is a required field for the splunk sink. We could work around this by defining a default value for the field (i.e. `false`).

```go
"splunk_secure": {
	Type:     schema.TypeBool,
	Optional: true,
	Default:  false,
}
```

@abulford I've tried reproducing some of the issues that were reported by adding some updating tests. I'll post them as suggestions on a review for more context.

alexkappa commented on auth0/resource_auth0_log_stream.go:
```diff
@@ -0,0 +1,346 @@
+						},
+						"splunk_secure": {
+							Type:     schema.TypeBool,
+							Optional: true,
```
```suggestion
							Optional: true,
							Default:  false,
```

alexkappa commented on auth0/resource_auth0_log_stream.go:
```diff
@@ -0,0 +1,346 @@
+	if err := api.LogStream.Create(c); err != nil {
+		return err
+	}
+	d.SetId(auth0.StringValue(c.ID))
```
```suggestion
	d.SetId(c.GetID())

	// The Management API only allows updating a log stream's status. Therefore
	// if the status field was present in the configuration, we perform an
	// additional operation to modify it.
	s := String(d, "status")
	if s != nil && s != c.Status {
		err := api.LogStream.Update(c.GetID(), &management.LogStream{Status: s})
		if err != nil {
			return err
		}
	}
```

alexkappa commented on auth0/resource_auth0_log_stream.go:
```diff
@@ -0,0 +1,346 @@
+		return err
+	}
+
+	d.SetId(auth0.StringValue(c.ID))
```
```suggestion
	d.SetId(c.GetID())
```

alexkappa commented on auth0/resource_auth0_log_stream.go:
```diff
@@ -0,0 +1,346 @@
+func expandLogStream(d ResourceData) *management.LogStream {
+
+	ls := &management.LogStream{
+		Name:   String(d, "name", IsNewResource()),
+		Type:   String(d, "type", IsNewResource()),
+		Status: String(d, "status"),
```
The `name` field is allowed to update apparently. See the [api docs](https://auth0.com/docs/api/management/v2#!/Log_Streams/patch_log_streams_by_id).

The `status` is a first for this provider. It can only be updated, not created. See also the suggestion on `createLogStream`.

```suggestion
		Name:   String(d, "name"),
		Type:   String(d, "type", IsNewResource()),
		Status: String(d, "status", Not(IsNewResource())),
```

alexkappa replied:
`Not` doesn't yet exist. But it's a trivial condition we can add to `resource_data.go`

```go
// Not is a condition that evaluates to true if its child condition evaluates to
// false. False otherwise.
func Not(condition Condition) Condition {
	return func(d ResourceData, key string) bool {
		return !condition.Eval(d, key)
	}
}
```

alexkappa commented on auth0/resource_auth0_log_stream.go:
```diff
@@ -0,0 +1,346 @@
+	List(d, "sink").Elem(func(d ResourceData) {
+		switch s {
+		case management.LogStreamTypeAmazonEventBridge:
+			ls.Sink = expandLogStreamSinkAmazonEventBridge(d)
```
> Note: For log streams of type eventbridge and eventgrid, updating the sink is not permitted.

From the [api docs](https://auth0.com/docs/api/management/v2#!/Log_Streams/patch_log_streams_by_id).

```suggestion
			if d.IsNewResource() {
				ls.Sink = expandLogStreamSinkAmazonEventBridge(d)
			}
```

alexkappa commented on auth0/resource_auth0_log_stream.go:
```diff
@@ -0,0 +1,346 @@
+		case management.LogStreamTypeAmazonEventBridge:
+			ls.Sink = expandLogStreamSinkAmazonEventBridge(d)
+		case management.LogStreamTypeAzureEventGrid:
+			ls.Sink = expandLogStreamSinkAzureEventGrid(d)
```
```suggestion
			if d.IsNewResource() {
				ls.Sink = expandLogStreamSinkAzureEventGrid(d)
			}
```

alexkappa commented on auth0/resource_auth0_log_stream.go:
```diff
@@ -0,0 +1,346 @@
+	o := &management.LogStreamSinkAmazonEventBridge{
+		AccountID:          String(d, "aws_account_id"),
+		Region:             String(d, "aws_region"),
+		PartnerEventSource: String(d, "aws_partner_event_source"),
```
This is a read-only field

```suggestion
```

alexkappa replied:
@mcalster should we remove this?

alexkappa commented on auth0/resource_auth0_log_stream_test.go:
```diff
@@ -0,0 +1,213 @@
+		},
+		Steps: []resource.TestStep{
+			{
+				Config: random.Template(logStreamHTTPConfig, rand),
```
Small naming change to be aligned with the codebase.
```suggestion
				Config: random.Template(testAccLogStreamHttpConfig, rand),
```

alexkappa commented on auth0/resource_auth0_log_stream_test.go:
```diff
@@ -0,0 +1,213 @@
+				Check: resource.ComposeTestCheckFunc(
+					random.TestCheckResourceAttr("auth0_log_stream.my_log_stream", "name", "Acceptance-Test-LogStream-http-{{.random}}", rand),
+					resource.TestCheckResourceAttr("auth0_log_stream.my_log_stream", "type", "http"),
+					//resource.TestCheckResourceAttr("auth0_log_stream.my_log_stream", "status", "active"),
```
```suggestion
```

alexkappa commented on auth0/resource_auth0_log_stream_test.go:
```diff
@@ -0,0 +1,213 @@
+					resource.TestCheckResourceAttr("auth0_log_stream.my_log_stream", "sink.0.http_content_format", "JSONLINES"),
+					resource.TestCheckResourceAttr("auth0_log_stream.my_log_stream", "sink.0.http_authorization", "AKIAXXXXXXXXXXXXXXXX"),
+				),
+			},
```
Additional tests to verify updates work as expected.

```suggestion
			},
			{
				Config: random.Template(testAccLogStreamHttpConfigUpdate, rand),
				Check: resource.ComposeTestCheckFunc(
					random.TestCheckResourceAttr("auth0_log_stream.my_log_stream", "name", "Acceptance-Test-LogStream-http-{{.random}}", rand),
					resource.TestCheckResourceAttr("auth0_log_stream.my_log_stream", "type", "http"),
					resource.TestCheckResourceAttr("auth0_log_stream.my_log_stream", "sink.0.http_endpoint", "https://example.com/webhook/logs"),
					resource.TestCheckResourceAttr("auth0_log_stream.my_log_stream", "sink.0.http_content_type", "application/json; charset=utf-8"),
					resource.TestCheckResourceAttr("auth0_log_stream.my_log_stream", "sink.0.http_content_format", "JSONARRAY"),
					resource.TestCheckResourceAttr("auth0_log_stream.my_log_stream", "sink.0.http_authorization", "AKIAXXXXXXXXXXXXXXXX"),
				),
			},
```

alexkappa commented on auth0/resource_auth0_log_stream_test.go:
```diff
@@ -0,0 +1,213 @@
+	})
+}
+
+const logStreamHTTPConfig = `
```
```suggestion
const testAccLogStreamHttpConfig = `
```

alexkappa commented on auth0/resource_auth0_log_stream_test.go:
```diff
@@ -0,0 +1,213 @@
+	}
+}
+`
+
```
```suggestion
const testAccLogStreamHttpConfigUpdate = `
resource "auth0_log_stream" "my_log_stream" {
	name = "Acceptance-Test-LogStream-http-{{.random}}"
	type = "http"
	sink {
	  http_endpoint = "https://example.com/webhook/logs"
	  http_content_type = "application/json; charset=utf-8"
	  http_content_format = "JSONARRAY"
	  http_authorization = "AKIAXXXXXXXXXXXXXXXX"
	}
}
`
```

mcalster commented:
> @abulford I've updated the test cases and it seems like I can reproduce your issue when changing the name of the log stream.
> Running with debug logging I get the following warning:
> 2020/11/13 20:49:46 [WARN] Provider "auth0" produced an unexpected new value for auth0_log_stream.my_log_stream, but we are tolerating it because it is using the legacy plugin SDK.
> The following problems may be the cause of any confusing errors from downstream operations:
> - .sink[0].splunk_secure: was null, but now cty.False
> - .sink[0].splunk_token: inconsistent values for sensitive attribute
> - .sink[0].splunk_domain: was null, but now cty.StringVal("")
> - .sink[0].azure_resource_group: was null, but now cty.StringVal("")
> - .sink[0].azure_subscription_id: inconsistent values for sensitive attribute
> - .sink[0].datadog_region: inconsistent values for sensitive attribute
> - .sink[0].datadog_api_key: inconsistent values for sensitive attribute
> - .sink[0].splunk_port: was null, but now cty.StringVal("")
> - .sink[0].aws_account_id: inconsistent values for sensitive attribute
> - .sink[0].aws_region: was null, but now cty.StringVal("")
> - .sink[0].azure_region: was null, but now cty.StringVal("")
> 
> I think the culprit is splunk_secure, but I'm unsure how to fix it :-/
> One way could be to change the type to string in the schema (and add a validator)and then convert back and forth when creating the structs? Input is appreciated 🧐
@abulford 
I think I found the cause to the issue. 
"For log streams of type eventbridge and eventgrid, updating the sink is not permitted." 
https://auth0.com/docs/api/management/v2#!/Log_Streams/patch_log_streams_by_id

@alexkappa how should this be implemented? It seems the all changes should result in a new change in this case then? By setting ForceNew on Azure and Amazon sink values?

mcalster commented:
@alexkappa I got a bit lost in the changes you made when just looking in the github interface. But now I think I got them all.

@abulford  I think the issue you experienced should be solved now, Azure and Amazon Sinks will trigger a new Resource when changed. (Updates are not supported for these two in the API)

abulford commented:
>I've changed the sensitivity as requested :-)

@mcalster thanks :)


>@abulford I think the issue you experienced should be solved now, Azure and Amazon Sinks will trigger a new Resource when changed. (Updates are not supported for these two in the API)

Thank you!  I've tried this out and updates do indeed seem to be fixed - I can update the name of the log stream and the update takes place successfully, I can update the AWS region (not tried account ID but expect it to behave the same) and the re-creation is successful, too.

I don't know how important it is, but something does still seem to be up with the other properties.
Here's how it looks when I create the resource:
```
  # auth0_log_stream.log_stream will be created
  + resource "auth0_log_stream" "log_stream" {
      + id     = (known after apply)
      + name   = "Auth0Events"
      + status = (known after apply)
      + type   = "eventbridge"

      + sink {
          + aws_account_id           = (sensitive value)
          + aws_partner_event_source = (sensitive value)
          + aws_region               = "us-east-1"
          + azure_partner_topic      = (sensitive value)
          + splunk_secure            = false
        }
    }
```
All looks good, though `azure_partner_topic` and `splunk_secure` are unexpectedly shown.

Once created, the state looks like this:
```
{
  "mode": "managed",
  "type": "auth0_log_stream",
  "name": "log_stream",
  "provider": "provider.auth0",
  "instances": [
    {
      "schema_version": 0,
      "attributes": {
        "id": "<REDACTED>",
        "name": "Auth0Events",
        "sink": [
          {
            "aws_account_id": "<REDACTED>",
            "aws_partner_event_source": "<REDACTED>",
            "aws_region": "us-east-1",
            "azure_partner_topic": "",
            "azure_region": "",
            "azure_resource_group": "",
            "azure_subscription_id": "",
            "datadog_api_key": "",
            "datadog_region": "",
            "http_authorization": "",
            "http_content_format": "",
            "http_content_type": "",
            "http_custom_headers": null,
            "http_endpoint": "",
            "splunk_domain": "",
            "splunk_port": "",
            "splunk_secure": false,
            "splunk_token": ""
          }
        ],
        "status": "active",
        "type": "eventbridge"
      },
      "private": "<REDACTED>"
    }
  ]
}
```

Changing the name causes an update, as expected:
```
  # auth0_log_stream.log_stream will be updated in-place
  ~ resource "auth0_log_stream" "log_stream" {
        id     = "<REDACTED>"
      ~ name   = "Auth0Events" -> "Auth0Events1"
        status = "active"
        type   = "eventbridge"

        sink {
            aws_account_id           = (sensitive value)
            aws_partner_event_source = (sensitive value)
            aws_region               = "us-east-1"
            http_custom_headers      = []
            splunk_secure            = false
        }
    }
```
We're now seeing `http_custom_headers` as well as `splunk_secure`.

Looking at the updated state, as well as the `name` having changed as we'd expect, the `http_custom_headers` is now set to `[]` instead of `null`:
```
11c11
<         "name": "Auth0Events",
---
>         "name": "Auth0Events1",
26c26
<             "http_custom_headers": null,
---
>             "http_custom_headers": [],
```

Changing the region causes a re-create, as we'd expect:
```
  # auth0_log_stream.log_stream must be replaced
-/+ resource "auth0_log_stream" "log_stream" {
      ~ id     = "<REDACTED>" -> (known after apply)
        name   = "Auth0Events1"
      ~ status = "active" -> (known after apply)
        type   = "eventbridge"

      ~ sink {
            aws_account_id           = (sensitive value)
          ~ aws_partner_event_source = (sensitive value)
          ~ aws_region               = "us-east-1" -> "eu-west-1" # forces replacement
          + azure_partner_topic      = (sensitive value)
          - http_custom_headers      = [] -> null
            splunk_secure            = false
        }
    }
```
Though again the plan is unexpectedly showing `azure_partner_topic`, `http_custom_headers` and `splunk_secure`.

Finally, destroying the resource works fine:
```
  # auth0_log_stream.log_stream will be destroyed
  - resource "auth0_log_stream" "log_stream" {
      - id     = "<REDACTED>" -> null
      - name   = "Auth0Events1" -> null
      - status = "active" -> null
      - type   = "eventbridge" -> null

      - sink {
          - aws_account_id           = (sensitive value)
          - aws_partner_event_source = (sensitive value)
          - aws_region               = "eu-west-1" -> null
          - http_custom_headers      = [] -> null
          - splunk_secure            = false -> null
        }
    }
```

I don't know if this causes any more harm than a bit of confusion when looking at the diffs, so maybe not worth holding up the PR for, but I thought it worth documenting the behaviour I'm currently seeing.  While this seems fine with the AWS EventBridge type, I don't know if it could cause issues with other types.


>We could work around this by defining a default value for the field (i.e. false).

@alexkappa false feels like an unsafe default for this property, given that this tells Auth0 whether to verify TLS or not.  If we can't validate this easily on the client side it might be preferable to let Auth0 return an error if the user hasn't set this.  Alternatively could it default to true?  Or is there any way we can avoid setting a default?

Another option could be to structure the schema a bit differently, so there are `eventbridge_sink`, `eventgrid_sink`, etc blocks rather than a single `sink` block.  This way those blocks could conflict with each other but within the blocks all the properties could be set as required or optional as necessary, if I'm understanding things right.

mcalster commented:
> All looks good, though `azure_partner_topic` and `splunk_secure` are unexpectedly shown.

I think the splunk_secure appears because of the default value of a Boolean.
I tried changing splunk_secure to a string that only accepts true/false. That seems to clean up the state. But I don't really like that solution. Another approach is to use a pointer but that doesn't make sense in the schema. Any suggestions are welcome.  

The azure_partner_topic is weird though, I will try to look into that.

abulford commented:
>Another approach is to use a pointer but that doesn't make sense in the schema.

Yeah, I think ultimately this would ideally be stored as a pointer to a bool, so it could be `nil`, `true` or `false`, but like you say, that doesn't seem supported by the schema.

I'm experimenting with the alternative format and it seems to work quite nicely, I'll submit a PR shortly to show you what I mean.

abulford commented:
@mcalster and @alexkappa I've created a [PR](https://github.com/mcalster/terraform-provider-auth0/pull/5) with the modified layout I was proposing, I've put the detail in there so won't repeat it all here.

It's closer than my suggestion to flatten all the properties, but admittedly still not completely consistent with the rest of the provider.  It does seem to bring a lot of benefit in validation and simplifying the situation with these unexpected properties.  If you're happy to go ahead I'll try to solve the not so nice situation where the `type` property doesn't need to match the sink.

alexkappa commented:
> @alexkappa how should this be implemented? It seems the all changes should result in a new change in this case then? By setting ForceNew on Azure and Amazon sink values?

That seems like a reasonable thing to do.

> @alexkappa false feels like an unsafe default for this property, given that this tells Auth0 whether to verify TLS or not. If we can't validate this easily on the client side it might be preferable to let Auth0 return an error if the user hasn't set this. Alternatively could it default to true? Or is there any way we can avoid setting a default?

Thats a fair point on having a default value or not. Having a default may not be the best idea as it will have the side effect of showing up when we don't need it.

Perhaps `RequiredWith` would be better in the way that we can mark it as required conditionally?

> @mcalster and @alexkappa I've created a PR with the modified layout I was proposing, I've put the detail in there so won't repeat it all here.

I would prefer to keep things consistent as much as possible. I do appreciate the power it gives us to define schema per sink type, but its inconsistent with the provider at this stage. Thanks for suggesting it though, as this approach hasn't crossed my mind before.

alexkappa commented on auth0/resource_auth0_log_stream.go:
```diff
@@ -0,0 +1,366 @@
+
+func createLogStream(d *schema.ResourceData, m interface{}) error {
+	ls := expandLogStream(d)
+	fmt.Printf("CREATE: ---------------> %+v\n", ls)
```
@mcalster consider using `TF_LOG=debug AUTH0_DEBUG=true` when running your tests instead of adding printf statements in your code.

alexkappa commented on go.mod:
```diff
@@ -5,5 +5,6 @@ go 1.13
 require (
 	github.com/hashicorp/go-multierror v1.1.0
 	github.com/hashicorp/terraform-plugin-sdk v1.16.0
+	gopkg.in/auth0.v4 v4.7.0
```
Something must have been mis-imported?

alexkappa commented on auth0/resource_auth0_log_stream.go:
```diff
@@ -0,0 +1,366 @@
+func expandLogStream(d ResourceData) *management.LogStream {
+
+	ls := &management.LogStream{
+		Name:   String(d, "name", HasChange()),
+		Type:   String(d, "type", HasChange()),
```
Probably better to use `IsNewResource` as `Type` isn't allowed to change. Also `Name` can be present any time.

```suggestion
		Name:   String(d, "name"),
		Type:   String(d, "type", IsNewResource()),
```

abulford commented:
>Thats a fair point on having a default value or not. Having a default may not be the best idea as it will have the side effect of showing up when we don't need it.

It seems that this property is showing up in the state whether there's a default value or not, and I don't know if setting validation will make a difference either, though I don't think I'll get chance to experiment with that today.  Ideally no default would be set (though we'll end up with `false` in the state) in the case that the property isn't going to be relevant, but in the case that it is relevant, the default would be `true`.  Could `DefaultFunc` be used for this?  If not then maybe just defaulting to `true` would be OK, to prevent accidentally disabling TLS verification when using the log stream with Splunk.

>Perhaps RequiredWith would be better in the way that we can mark it as required conditionally?
>
>Edit: here's an example from the aws provider doing something similar.

Ooh, I was about to reply to say that I'd tried `RequiredWith` for some other properties but it didn't seem to work within the nested structure, but I had been trying `sink.aws_account_id` rather than `sink.0.aws_account_id`.  That opens up more validation options than I'd realised were possible with the existing structure 👍.

>I would prefer to keep things consistent as much as possible. I do appreciate the power it gives us to define schema per sink type, but its inconsistent with the provider at this stage. Thanks for suggesting it though, as this approach hasn't crossed my mind before.

Fair enough :), I agree it would be a departure from the existing convention.  It might be worth considering something like this if there's any plan to refactor for a major release as I think it could improve usability and address some state migration issues I've found previously, but yeah, this PR probably isn't the place for it!

mcalster commented:
@alexkappa I've added RequiredWith to the sink values, and removed default value from the slunk_secure

abulford commented:
@mcalster I've tested your latest changes and all looks good!  I created, updated, recreated and deleted an AWS EventBridge resource and it all worked fine.  The `splunk_secure` and `http_custom_headers` properties are still showing up but it doesn't actually break anything.

I also tried removing the `aws_region` property and validation correctly picked that up.  I tried adding an unexpected property, `aws_partner_event_source`, and validation didn't pick that up, so Auth0 ends up returning a 400 Bad request, with `Payload validation error: 'Data does not match any schemas from 'oneOf''`.  This could be addressed with `ConflictsWith`, or maybe some fancy validation checking the `type` property, but maybe this extra validation could be added in a later PR.

I'm also unsure if all the `RequiredWith` constraints are necessary in the HTTP and Splunk streams - are some of those optional properties in the Auth0 management API?  But again maybe not worth worrying about now as it's a minor detail.

alexkappa commented:
@mcalster thanks for making the changes, I would say this is good to merge. Last thing before I do that, why did you decide to define `aws_account_id`, `aws_partner_event_source`, `azure_subscription_id`, `azure_partner_topic` and `datadog_region` as sensitive? If I understand correctly these aren't particularly sensitive values right?

mcalster commented:
I started out with having them as non sensitive, but got some review
comments on that approach.

It will be fine by me to remove it.

man. 16. nov. 2020 kl. 14.06 skrev Alex Kalyvitis <notifications@github.com
>:

> @mcalster <https://github.com/mcalster> thanks for making the changes, I
> would say this is good to merge. Last thing before I do that, why did you
> decide to define aws_account_id, aws_partner_event_source,
> azure_subscription_id, azure_partner_topic and datadog_region as
> sensitive? If I understand correctly these aren't particularly sensitive
> values right?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/alexkappa/terraform-provider-auth0/pull/270#issuecomment-727977619>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAX7II5KQ4DZW3MB4VDHL4LSQEPURANCNFSM4R3ZKSOQ>
> .
>
-- 
-- Venlig Hilsen
Kim Alster

alexkappa commented:
Great, I will make the change and merge it in then!

abulford commented:
>why did you decide to define aws_account_id, aws_partner_event_source, azure_subscription_id, azure_partner_topic and datadog_region as sensitive? If I understand correctly these aren't particularly sensitive values right?

>I started out with having them as non sensitive, but got some review comments on that approach.

My feedback was to say that I wasn't sure that the `aws_account_id` or `aws_region` should be sensitive.  On the basis that the `aws_account_id` remained marked as sensitive I suggested erring on the side of sensitive for the `aws_partner_event_source` and `azure_partner_topic` properties as well, but not especially convinced either way.  I'm sorry if I didn't explain myself properly.

Would certainly be happy for all those properties mentioned to not be marked as sensitive.
